Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/GetExamUseCase.java

```diff
@@ -15,25 +16,29 @@
  * @author - johnny850807@gmail.com (Waterball)
  */
 @Named
-@AllArgsConstructor
-public class GetExamUseCase {
-    private final ExamRepository examRepository;
+public class GetExamUseCase extends AbstractExamUseCase {
+
+    public GetExamUseCase(ExamRepository examRepository) {
+        super(examRepository);
+    }
 
     public void execute(Request request, ExamPresenter presenter) throws ExamineeOnlyOperationException {
         Exam exam = findExam(request.examId);
         onlyExamineeCanAccessTheExam(request, exam);
+        onlyWhitelistIpAddressExamineeCanAccessTheOngoingExam(request, exam);
         presenter.showExam(exam);
     }
 
     private void onlyExamineeCanAccessTheExam(Request request, Exam exam) {
-        if (request.isOnlyExamineeCanAccess() && !exam.hasExaminee(request.studentId)) {
+        if (request.onlyExamineeCanAccess && !exam.hasExaminee(request.studentId)) {
             throw new ExamineeOnlyOperationException();
         }
     }
 
-    private Exam findExam(int examId) {
-        return examRepository.findById(examId)
-                .orElseThrow(() -> notFound(Exam.class).id(examId));
+    private void onlyWhitelistIpAddressExamineeCanAccessTheOngoingExam(Request request, Exam exam) {
+        if (request.onlyExamineeCanAccess && exam.isOngoing() && !exam.hasIpAddress(new IpAddress(request.ipAddress))) {
+            throw notFound(Exam.class).id(request.examId);
+        }
```

Comment:
We have checked the examinee's identity in `onlyExamineeCanAccessTheExam()`.
I think we don't need to check it again.

You can change `onlyWhitelistIpAddressExamineeCanAccessTheOngoingExam` as follows, 
```java
private void onlyWhitelistIpAddresCanAccessOngoingExam(Request request, Exam exam) {
        if (exam.isOngoing() && !exam.hasIpAddress(new IpAddress(request.ipAddress))) {
            throw notFound(Exam.class).id(request.examId);
        }
```

Reply:
No, we can't, because ```onlyWhitelistIpAddresCanAccessOngoingExam ``` only block the examinee whose IP is not in the exam white list, if I remove the examinee validation, it will also block admin.